pulsar-broker/src/main/java/org/apache/pulsar/broker/service/DrainingHashesTracker.java

```diff
@@ -0,0 +1,250 @@
+            return "DrainingHashEntry(consumer=" + this.consumer + ", refCount=" + this.refCount + ", blockedCount="
+                    + this.blockedCount + ")";
+        }
+    }
+
+    /**
+     * Interface for handling the unblocking of sticky key hashes.
+     */
+    public interface UnblockingHandler {
+        /**
+         * Handle the unblocking of a sticky key hash.
+         *
+         * @param stickyKeyHash the sticky key hash that has been unblocked, or -1 if hash unblocking is done in batch
+         */
+        void stickyKeyHashUnblocked(int stickyKeyHash);
+    }
+
+    public DrainingHashesTracker(String dispatcherName, UnblockingHandler unblockingHandler) {
+        this.dispatcherName = dispatcherName;
+        this.unblockingHandler = unblockingHandler;
+    }
+
+    /**
+     * Add an entry to the draining hashes tracker.
+     *
+     * @param consumer the consumer
+     * @param stickyHash the sticky hash
+     */
+    public synchronized void addEntry(Consumer consumer, int stickyHash) {
+        if (stickyHash == 0) {
+            throw new IllegalArgumentException("Sticky hash cannot be 0");
+        }
+        DrainingHashEntry entry = drainingHashes.get(stickyHash);
+        if (entry == null) {
+            entry = new DrainingHashEntry(consumer);
+            drainingHashes.put(stickyHash, entry);
+        } else if (entry.getConsumer() != consumer) {
+            log.error("[{}] Consumer {} is already draining hash {}. Same hash being used for consumer {}. "
+                            + "This call is ignored since there's a problem in consistency.",
+                    dispatcherName,
+                    entry.getConsumer(), stickyHash, consumer);
+            return;
+        }
+        entry.incrementRefCount();
+    }
+
+    /**
+     * Start a batch operation. There could be multiple nested batch operations.
+     * The unblocking of sticky key hashes will be done only when the last batch operation ends.
+     */
+    public synchronized void startBatch() {
+        batchLevel++;
+    }
+
+    /**
+     * End a batch operation.
+     */
+    public synchronized void endBatch() {
+        if (--batchLevel == 0 && unblockedWhileBatching) {
+            unblockedWhileBatching = false;
+            unblockingHandler.stickyKeyHashUnblocked(-1);
+        }
+    }
+
+    /**
+     * Reduce the reference count for a given sticky hash.
+     *
+     * @param consumer   the consumer
+     * @param stickyHash the sticky hash
+     * @param closing
+     */
+    public synchronized void reduceRefCount(Consumer consumer, int stickyHash, boolean closing) {
+        if (stickyHash == 0) {
+            return;
+        }
+        DrainingHashEntry entry = drainingHashes.get(stickyHash);
+        if (entry == null) {
+            return;
+        }
+        if (entry.getConsumer() != consumer) {
+            log.error("[{}] Consumer {} is already draining hash {}. Same hash is being acknowledged by consumer {}."
```

Comment:
Thanks for the suggestion. I ended up using IllegalStateException since if this would happen some of the current assumptions is failing and it's better to catch the condition early.